lib/compatibility.bzl

```diff
@@ -0,0 +1,185 @@
+"""Skylib module of convenience functions for `target_compatible_with`.
+
+Load the macros as follows in your `BUILD` files:
+```python
+load("@bazel_skylib//lib:compatibility.bzl", "compatibility")
+```
+
+See the [Platform docs](https://bazel.build/docs/platforms#skipping-incompatible-targets) for
+more information.
+"""
+
+load(":selects.bzl", "selects")
+
+def _get_name_from_target_list(targets, joiner = " or "):
+    """Join a list of strings into a string which is suitable as a target name.
+
+    The return value has a hash appended so that it is different between multiple "targets" values
+    that have the same name portion. For readability, we keep only the name portion of the
+    specified targets. But for uniqueness we need the hash. For example, the following two calls
+    may return strings as follows. The hashes are arbitrary in this example.
+
+        >>> _get_name_from_target_list(["@platforms//os:linux", "@platforms//cpu:x86_64"])
+        "linux or x86_64 (3ef2349)"
+        >>> _get_name_from_target_list(["//some/custom:linux", "//some/custom:x86_64"])
+        "linux or x86_64 (98ab64)"
+
+    Args:
+      targets: A list of target names.
+      joiner: An optional string to use for joining the list.
+
+    Returns:
+      A string which is a valid target name.
+    """
+
+    # Compute absolute labels from the user-specified ones. They can be
+    # relative or absolute. We do this to make the final name as unique as
+    # possible via a hash.
+    package_label = Label("//" + native.package_name())
+    absolute_targets = sorted([package_label.relative(target) for target in targets])
+    joined_names = joiner.join([target.name for target in absolute_targets])
+    name = "%s (%x)" % (joined_names, hash(str(absolute_targets)))
+    return name
+
+def _maybe_make_unique_incompatible_value(name):
+    """Creates a `native.constraint_value` which is "incompatible."
+
+    When composing selects which could all resolve to "incompatible" we need distinct labels.
+    This will create a `constraint_value` with the given name, if it does not already exist.
+
+    Args:
+      name: A target name to check and use.
+    """
+    if not native.existing_rule(name):
+        native.constraint_value(
+            name = name,
+            constraint_setting = "@platforms//:incompatible_setting",
+        )
```

Comment:
There already is a `constraint_value`: `@platforms//:incompatible`

Reply:
I ended up implementing the approach I thought of in the above comment. Please let me know what you think. Happy to add more comments etc. if you approve of the approach.